File: src/types/lesson.ts
/**
 * Lesson Data Model for Learn to Read, Read to Learn
 *
 * This module defines the TypeScript interfaces for lessons, activities,
 * and progress tracking in the L2RR2L learning system.
 */

// ============================================================================
// Constants (using const objects for erasableSyntaxOnly compatibility)
// ============================================================================

/** Subject categories for lessons */
export const LessonSubject = {
  PHONICS: 'phonics',
  SPELLING: 'spelling',
  SIGHT_WORDS: 'sight-words',
  READING: 'reading',
  WORD_FAMILIES: 'word-families',
  VOCABULARY: 'vocabulary',
  COMPREHENSION: 'comprehension',
} as const

export type LessonSubject = typeof LessonSubject[keyof typeof LessonSubject]

/** Difficulty levels for lessons */
export const DifficultyLevel = {
  BEGINNER: 'beginner',
  INTERMEDIATE: 'intermediate',
  ADVANCED: 'advanced',
} as const

export type DifficultyLevel = typeof DifficultyLevel[keyof typeof DifficultyLevel]

/** Types of activities within a lesson */
export const ActivityType = {
  /** Read text aloud or silently */
  READING: 'reading',
  /** Spell words by arranging letters */
  SPELLING: 'spelling',
  /** Practice letter sounds and blends */
  PHONICS: 'phonics',
  /** Recognize and memorize sight words */
  SIGHT_WORDS: 'sight-words',
  /** Answer questions about content */
  QUIZ: 'quiz',
  /** Match related items (words to pictures, rhymes, etc.) */
  MATCHING: 'matching',
  /** Fill in missing letters or words */
  FILL_IN_BLANK: 'fill-in-blank',
  /** Listen and repeat words/phrases */
  LISTEN_REPEAT: 'listen-repeat',
  /** Build words from word families */
  WORD_BUILDING: 'word-building',
} as const

export type ActivityType = typeof ActivityType[keyof typeof ActivityType]

/** Status of a lesson for a student */
export const LessonStatus = {
  NOT_STARTED: 'not-started',
  IN_PROGRESS: 'in-progress',
  COMPLETED: 'completed',
} as const

export type LessonStatus = typeof LessonStatus[keyof typeof LessonStatus]

// ============================================================================
// Activity Interfaces
// ============================================================================

/** Base interface for all activity types */
export interface BaseActivity {
  id: string
  type: ActivityType
  /** Instructions displayed to the student */
  instructions: string
  /** Text to be spoken by TTS (optional, defaults to instructions) */
  spokenInstructions?: string
  /** Order within the lesson (0-indexed) */
  order: number
  /** Points awarded for completing this activity */
  points?: number
}

/** Reading activity - display text for reading */
export interface ReadingActivity extends BaseActivity {
  type: 'reading'
  /** Text content to read */
  content: string
  /** Optional image to accompany the text */
  imageUrl?: string
  /** Whether to use TTS to read the text aloud */
  readAloud?: boolean
}

/** Spelling activity - spell a word from scrambled letters */
export interface SpellingActivity extends BaseActivity {
  type: 'spelling'
  /** Word to spell */
  word: string
  /** Hint image or emoji */
  hint?: string
  /** Optional audio pronunciation */
  audioUrl?: string
}

/** Phonics activity - practice sounds */
export interface PhonicsActivity extends BaseActivity {
  type: 'phonics'
  /** The sound/phoneme being practiced (e.g., "sh", "th", "a") */
  sound: string
  /** Example words containing this sound */
  exampleWords: string[]
  /** Position of sound in words: beginning, middle, end */
  soundPosition?: 'beginning' | 'middle' | 'end' | 'any'
}

/** Sight words activity - recognize common words */
export interface SightWordsActivity extends BaseActivity {
  type: 'sight-words'
  /** Words to practice */
  words: string[]
  /** Whether to show words in sentences */
  showInContext?: boolean
}

/** Quiz activity - answer questions */
export interface QuizActivity extends BaseActivity {
  type: 'quiz'
  /** The question to answer */
  question: string
  /** Available answer options */
  options: string[]
  /** Index of correct answer in options array */
  correctIndex: number
  /** Explanation shown after answering */
  explanation?: string
}

/** Matching activity - match pairs */
export interface MatchingActivity extends BaseActivity {
  type: 'matching'
  /** Pairs to match [item1, item2] */
  pairs: [string, string][]
  /** Type of matching (word-picture, word-word, etc.) */
  matchType: 'word-picture' | 'word-word' | 'word-definition' | 'rhyme'
}

/** Fill in the blank activity */
export interface FillInBlankActivity extends BaseActivity {
  type: 'fill-in-blank'
  /** Sentence with ___ for blank */
  sentence: string
  /** The correct answer */
  answer: string
  /** Optional word bank */
  wordBank?: string[]
}

/** Listen and repeat activity */
export interface ListenRepeatActivity extends BaseActivity {
  type: 'listen-repeat'
  /** Word or phrase to repeat */
  phrase: string
  /** Whether to check pronunciation */
  checkPronunciation?: boolean
}

/** Word building activity */
export interface WordBuildingActivity extends BaseActivity {
  type: 'word-building'
  /** Word family pattern (e.g., "-at", "-an") */
  pattern: string
  /** Beginning sounds to combine with pattern */
  onsets: string[]
  /** Resulting words */
  words: string[]
}

/** Union type for all activity types */
export type LessonActivity =
  | ReadingActivity
  | SpellingActivity
  | PhonicsActivity
  | SightWordsActivity
  | QuizActivity
  | MatchingActivity
  | FillInBlankActivity
  | ListenRepeatActivity
  | WordBuildingActivity

// ============================================================================
// Lesson Interface
// ============================================================================

/** Main lesson interface */
export interface Lesson {
  /** Unique identifier */
  id: string
  /** Display title */
  title: string
  /** Brief description */
  description: string
  /** Subject category */
  subject: LessonSubject
  /** Difficulty level */
  difficulty: DifficultyLevel
  /** Learning objectives */
  objectives: string[]
  /** Ordered list of activities */
  activities: LessonActivity[]
  /** Estimated duration in minutes */
  durationMinutes: number
  /** Prerequisite lesson IDs */
  prerequisites?: string[]
  /** Tags for filtering/search */
  tags?: string[]
  /** Thumbnail image URL */
  thumbnailUrl?: string
  /** Target age range */
  ageRange?: {
    min: number
    max: number
  }
  /** Creation timestamp */
  createdAt: string
  /** Last update timestamp */
  updatedAt: string
}

// ============================================================================
// Progress Tracking
// ============================================================================

/** Progress on a single activity */
export interface ActivityProgress {
  activityId: string
  completed: boolean
  score?: number
  attempts: number
  timeSpentSeconds: number
  completedAt?: string
}

/** Progress on a lesson */
export interface LessonProgress {
  lessonId: string
  childId: string
  status: LessonStatus
  /** Current activity index (for resuming) */
  currentActivityIndex: number
  /** Progress on individual activities */
  activityProgress: ActivityProgress[]
  /** Overall score (percentage) */
  overallScore?: number
  /** Total time spent on lesson */
  totalTimeSeconds: number
  /** When lesson was started */
  startedAt: string
  /** When lesson was completed */
  completedAt?: string
}

// ============================================================================
// API Types
// ============================================================================

/** Request to create/update a lesson */
export interface LessonCreateRequest {
  title: string
  description: string
  subject: LessonSubject
  difficulty: DifficultyLevel
  objectives: string[]
  activities: Omit<LessonActivity, 'id'>[]
  durationMinutes: number
  prerequisites?: string[]
  tags?: string[]
  thumbnailUrl?: string
  ageRange?: { min: number; max: number }
}

/** Query parameters for listing lessons */
export interface LessonListQuery {
  subject?: LessonSubject
  difficulty?: DifficultyLevel
  tags?: string[]
  search?: string
  limit?: number
  offset?: number
}

/** Response for lesson list */
export interface LessonListResponse {
  lessons: Lesson[]
  total: number
  limit: number
  offset: number
}

// ============================================================================
// Helper Types
// ============================================================================

/** Lesson with progress information for a specific child */
export interface LessonWithProgress extends Lesson {
  progress?: LessonProgress
}

/** Summary statistics for a child's learning */
export interface LearningStats {
  childId: string
  totalLessonsStarted: number
  totalLessonsCompleted: number
  averageScore: number
  totalTimeMinutes: number
  subjectProgress: Record<LessonSubject, {
    lessonsCompleted: number
    averageScore: number
  }>
  recentActivity: {
    lessonId: string
    lessonTitle: string
    status: LessonStatus
    lastAccessedAt: string
  }[]
}

File: ios/L2RR2L/Core/Models/Lesson.swift
import Foundation

// MARK: - Lesson Enums

enum LessonSubject: String, Codable, CaseIterable {
    case phonics
    case spelling
    case sightWords = "sight-words"
    case reading
    case wordFamilies = "word-families"
    case vocabulary
    case comprehension
}

enum DifficultyLevel: String, Codable, CaseIterable {
    case beginner
    case intermediate
    case advanced
}

enum ActivityType: String, Codable, CaseIterable {
    case reading
    case spelling
    case phonics
    case sightWords = "sight-words"
    case quiz
    case matching
    case fillInBlank = "fill-in-blank"
    case listenRepeat = "listen-repeat"
    case wordBuilding = "word-building"
}

// MARK: - Activity Models

protocol LessonActivityProtocol: Codable, Identifiable {
    var id: String { get }
    var type: ActivityType { get }
    var instructions: String { get }
    var spokenInstructions: String? { get }
    var order: Int { get }
    var points: Int? { get }
}

struct ReadingActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let content: String
    let imageUrl: String?
    let readAloud: Bool?
}

struct SpellingActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let word: String
    let hint: String?
    let audioUrl: String?
}

struct PhonicsActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let sound: String
    let exampleWords: [String]
    let soundPosition: String?
}

struct SightWordsActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let words: [String]
    let showInContext: Bool?
}

struct QuizActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let question: String
    let options: [String]
    let correctIndex: Int
    let explanation: String?
}

struct MatchingActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let pairs: [[String]]
    let matchType: String
}

struct FillInBlankActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let sentence: String
    let answer: String
    let wordBank: [String]?
}

struct ListenRepeatActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let phrase: String
    let checkPronunciation: Bool?
}

struct WordBuildingActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?
    let pattern: String
    let onsets: [String]
    let words: [String]
}

// MARK: - Lesson Activity (Type-Erased)

struct LessonActivity: Codable, Identifiable, Equatable {
    let id: String
    let type: ActivityType
    let instructions: String
    let spokenInstructions: String?
    let order: Int
    let points: Int?

    // Activity-specific fields (optional based on type)
    let content: String?
    let imageUrl: String?
    let readAloud: Bool?
    let word: String?
    let hint: String?
    let audioUrl: String?
    let sound: String?
    let exampleWords: [String]?
    let soundPosition: String?
    let words: [String]?
    let showInContext: Bool?
    let question: String?
    let options: [String]?
    let correctIndex: Int?
    let explanation: String?
    let pairs: [[String]]?
    let matchType: String?
    let sentence: String?
    let answer: String?
    let wordBank: [String]?
    let phrase: String?
    let checkPronunciation: Bool?
    let pattern: String?
    let onsets: [String]?
}

// MARK: - Lesson

struct Lesson: Codable, Identifiable, Equatable {
    let id: String
    let title: String
    let description: String
    let subject: LessonSubject
    let difficulty: DifficultyLevel
    let objectives: [String]
    let activities: [LessonActivity]
    let durationMinutes: Int
    let prerequisites: [String]?
    let tags: [String]?
    let thumbnailUrl: String?
    let ageRange: AgeRange?
    let createdAt: String
    let updatedAt: String

    struct AgeRange: Codable, Equatable {
        let min: Int
        let max: Int
    }
}

// MARK: - Lesson List Response

struct LessonListResponse: Codable, Equatable {
    let lessons: [Lesson]
    let total: Int
    let limit: Int
    let offset: Int
}

// MARK: - Lesson Search Filters

struct LessonSearchFilters: Codable {
    var subject: LessonSubject?
    var difficulty: DifficultyLevel?
    var tags: [String]?
    var search: String?
    var limit: Int?
    var offset: Int?
}
